### PR TITLE
Setup Offline Mode on DocV and EnhancedDocV

### PR DESCRIPTION
**Pull request by jumaallan**
Story: https://app.shortcut.com/smileid/story/xxx

## Summary

A few sentences/bullet points about the changes

## Known Issues

Any shortcomings in your work. This may include corner cases not correctly handled or issues related
to but not within the scope of your PR. Design compromises should be discussed here if they were not
already discussed above.

## Test Instructions

Concise test instructions on how to verify that your feature works as intended. This should include
changes to the development environment (if applicable) and all commands needed to run your work.

## Screenshot

If applicable (e.g. UI changes), add screenshots to help explain your work.

### CODE REVIEW

File: Example/Podfile.lock [hunk 1]
```diff
@@ -39,4 +39,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: 81f36d854eaf9451aa489ca090d5c32269373e5d
 
-COCOAPODS: 1.13.0
+COCOAPODS: 1.14.3
```
Comment (vanshg) [inline Example/Podfile.lock:42]: CocoaPods 1.15.2 is the latest now

File: Example/SmileID.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -31,6 +31,7 @@
 		620F1E982B69194900185CD2 /* AlertView.swift in Sources */ = {isa = PBXBuildFile; fileRef = 620F1E972B69194900185CD2 /* AlertView.swift */; };
 		620F1E9A2B691ABB00185CD2 /* (null) in Resources */ = {isa = PBXBuildFile; };
 		624777D02B0CDC9F00952842 /* EnhancedKycWithIdInputScreen.swift in Sources */ = {isa = PBXBuildFile; fileRef = 624777CF2B0CDC9F00952842 /* EnhancedKycWithIdInputScreen.swift */; };
+		629BBF832B711AF600115769 /* smile_config.json in Resources */ = {isa = PBXBuildFile; fileRef = 629BBF822B711AF600115769 /* smile_config.json */; };
```
Comment (vanshg) [inline Example/SmileID.xcodeproj/project.pbxproj:34]: We should not add this file here

File: Sources/SmileID/Classes/DocumentVerification/Model/DocumentCaptureResultStore.swift [hunk 3]
```diff
@@ -2,8 +2,8 @@ import Foundation
 
 struct DocumentCaptureResultStore {
     var allFiles: [URL]
-    var documentFront: URL
+    var documentFront: URL?
```
Comment (vanshg) [inline Sources/SmileID/Classes/DocumentVerification/Model/DocumentCaptureResultStore.swift:5]: Why does this need to become optional?

File: Sources/SmileID/Classes/DocumentVerification/Model/OrchestratedDocumentVerificationViewModel.swift [hunk 4]
```diff
@@ -118,23 +120,16 @@ internal class IOrchestratedDocumentVerificationViewModel<T, U: JobResult>: Obse
...
+            front: documentFrontFile,
+            back: documentBackFile,
+            selfie: selfieFile,
+            livenessImages: livenessFiles,
+            countryCode: countryCode,
+            documentType: documentType,
+            jobId: jobId,
+            saveStage:.pending
+        ) else {
             return
```
Comment (vanshg) [inline Sources/SmileID/Classes/DocumentVerification/Model/OrchestratedDocumentVerificationViewModel.swift:133]: I would err on the side of making `saveDocumentImages` responsible for only saving images, and not for error handling. I think the error handling should be done here and `saveDocumentImages` should be marked as `throws`

File: Sources/SmileID/Classes/DocumentVerification/Model/OrchestratedDocumentVerificationViewModel.swift [hunk 5]
```diff
@@ -147,48 +142,99 @@ internal class IOrchestratedDocumentVerificationViewModel<T, U: JobResult>: Obse
...
+            },
+            receiveValue: { [weak self] response in
+                _ = self?.saveDocumentImages(
+                    front: documentFrontFile,
+                    back: self?.documentBackFile,
+                    selfie: selfieFile,
+                    livenessImages: self?.livenessFiles,
+                    countryCode: self?.countryCode ?? "",
+                    documentType: self?.documentType,
+                    jobId: self?.jobId ?? "", // TODO - cannot be null, will fix
```
Comment (vanshg) [inline Sources/SmileID/Classes/DocumentVerification/Model/OrchestratedDocumentVerificationViewModel.swift:196]: In this PR or next?

File: Sources/SmileID/Classes/Helpers/LocalStorage.swift [hunk 6]
```diff
@@ -2,7 +2,9 @@ import Foundation
 import Zip
 
 public class LocalStorage {
-    private static let defaultFolderName = "sid_jobs"
+    private static let defaultFolderName = "SmileID"
+    private static let pendingFolderName = "pending"
+    private static let completedFolderName = "completed"
```
Comment (vanshg) [inline Sources/SmileID/Classes/Helpers/LocalStorage.swift:7]: As we discussed today, we should start calling everything as submitted/unsubmitted rather than pending/completed

File: Sources/SmileID/Classes/Helpers/LocalStorage.swift [hunk 7]
```diff
@@ -20,12 +22,30 @@ public class LocalStorage {
         }
     }
 
+    static var pendingDirectory: URL {
```
Comment (vanshg) [inline Sources/SmileID/Classes/Helpers/LocalStorage.swift:25]: Any reason this can't just be a constant instead of a getter? i.e. something like `static let pendingDirectory = defaultDirectory.appendingPathComponent`

File: Sources/SmileID/Classes/Helpers/LocalStorage.swift [hunk 8]
```diff
@@ -20,12 +22,30 @@ public class LocalStorage {
...
+
+    enum SaveType {
+        case pending
+        case completed
+    }
+
+    /// todo - rework this logic
     static func saveImage(
         image: Data,
         to folder: String = "sid-\(UUID().uuidString)",
```
Comment (vanshg) [inline Sources/SmileID/Classes/Helpers/LocalStorage.swift:45]: Is it possible to remove this parameter right now since it has a default value? If not we can do in the later PRs as you indicated

File: Sources/SmileID/Classes/Helpers/LocalStorage.swift [hunk 9]
```diff
@@ -89,32 +125,49 @@ public class LocalStorage {
...
-    ///   - livenessImages: The selfie capture liveness images
     ///   - selfie: The selfie capture
+    ///   - livenessImages: The selfie capture liveness images
     ///   - countryCode: The document country code
     ///   - documentType: The optional document type
     ///   - folder: The name of the folder the files should be saved
     /// - Returns: A document result store which encapsulates the urls of the saved images
     static func saveDocumentImages(
-        front: Data,
+        front: Data?,
```
Comment (vanshg) [inline Sources/SmileID/Classes/Helpers/LocalStorage.swift:135]: When would the front document image be null?

File: Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift [hunk 10]
```diff
@@ -290,7 +290,8 @@ public class SelfieViewModel: ObservableObject, ARKitSmileDelegate {
                 }
                 let infoJson = try LocalStorage.createInfoJson(
                     selfie: selfieImage,
-                    livenessImages: livenessImages
+                    livenessImages: livenessImages,
+                    jobId: jobId
```
Comment (vanshg) [inline Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift:294]: nit: I think jobId should become the first parameter

File: Sources/SmileID/Classes/SmileID.swift [hunk 11]
```diff
@@ -71,6 +72,20 @@ public class SmileID {
...
+    /// - Parameter allowOfflineMode: A boolean to enable offline mode or not
+    public class func setAllowOfflineMode(allowOfflineMode: Bool) {
+        SmileID.allowOfflineMode = allowOfflineMode
+    }
+
+    public class func submitJob(jobId: String, deleteFilesOnSuccess: Bool = false) {}
+
+    public class func cleanUpJob(jobId: String) {}
+
+    public class func getUnsubmittedJobs() -> [String] {
```
Comment (vanshg) [inline Sources/SmileID/Classes/SmileID.swift:85]: updates as we discussed

**Review by vanshg — COMMENTED** [review]
This looks like a great start, thanks for making it to a new branch. Can you also add tests?

File: Sources/SmileID/Classes/Helpers/LocalStorage.swift (hunk 7 above)
Comment (jumaallan) [reply Sources/SmileID/Classes/Helpers/LocalStorage.swift:25]: It can `throw`, and we can't assign to a constant if it throws
